R/cr.R

```diff
@@ -194,3 +169,117 @@ modify_cr_data <- function(data, cr) {
+#' @param formula an object of class formula or a string convertible to it.
+#' The model formula corresponding to a gam or glm.
+#' @param family the family of the gam to be modelled. Only poisson is
+#'  reasonable. Hence, any other input will not be accepted.
+#' @param data A data.frame of class ped_cr that features time-to-event data
+#' and convariates. (see https://adibender.github.io/pammtools/)
+#' @param offset The offset for each observation. Contained in data.
+#' @param m_type A character indicating the type of model to be fit: either 
+#' "glm" or "gam".
+#' @param ... additional arguments passed to the gam/glm function.
+#' @return a list of gams or glms - one entry for a single competing risk.
+#' @importFrom mgcv gam
+#' @importFrom stats glm
+fit_cr <- function(formula, family, data, offset, m_type, ...) {
+  crs <- unique(data$ped_status)
+  crs <- crs[!(crs == 0)]
+  n_crs <- length(crs)
+  res <- vector(mode = "list", length = n_crs)
+  names(res) <- crs
+  for (i in 1:n_crs) {
+    # this function is supposed to make a ped_cr object to a ped object
+    # where we only investiagte one of the competing risks
+    current_data <- modify_cr_data(data, cr = crs[i])
+    command <- paste(m_type, "(formula = formula, family = family, ", 
+                    "data = current_data, offset = offset, ...)", sep = "")
+    res[[i]] <- eval(parse(text = command))
+  }
+  return(res)
+}
+
+#' Input checking for cr PEMs / PAMMs
+#' 
+#' This function fits a gams/glms depending on the input for a cr framework.
+#' It is the operative sub-routine of glm_cr and gam_cr.
+#' @param formula an object of class formula or a string convertible to it.
+#' The model formula corresponding to a gam or glm.
+#' @param data A data.frame of class ped_cr that features time-to-event data
+#' and convariates. (see https://adibender.github.io/pammtools/)
+#' @param offset The offset for each observation. Contained in data.
+#' @return An assertion if there is false input.
+#' @import checkmate
+check_input <- function(formula, data, offset) {
+  assert_formula(formula)
+  assert_data_frame(data)
+  if (is.null(offset)) stop("You need to specifiy an offset.")
+}
+
+
+make_numeric <- function(data, status_str, censor_code = 0) {
+  risks <- unique(data[[status_str]])
+  risks <- risks[risks != censor_code]
+  data[data[[status_str]] == censor_code, status_str] <- 0
+  for (i in 1:length(risks)) {
+    data[data[[status_str]] == risks[i], status_str] <- i
+  }
+  data[[status_str]] <- as.numeric(data[[status_str]])
+  attr(data, "risks") <- risks
+  data
+}
+
+hazard_adder_cr <- function(newdata, object, hazard_function, type, ci, se_mult, 
+                            ci_type, overwrite, time_var, ...) {
+  new_cols <- as.data.frame(matrix(0, nrow = nrow(newdata), 
+                                   ncol = length(object) + 
+                                     length(object) * ci * 2))
+  j <- 1
+  for (i in 1:length(object)) {
+    if (ci) {
```

Comment:
ich glaub das if-else hier ließe sich vermeiden wenn man einmal am anfang den index j bzw die indexmenge in die man zuweist je nach dem wert von ci generiert. das wäre dann kürzer und DRY